ddtrace/internal/_encoding.pyx

```diff
@@ -91,180 +91,223 @@ cdef class Packer(object):
+        cdef int ret
+        cdef dict d
 
-                if ret == 0:
-                    ret = pack_bytes(&self.pk, <char *>b"trace_id", 8)
-                    if ret != 0: return ret
-                    ret = self._pack(o.trace_id)
-                    if ret != 0: return ret
-
-                    ret = pack_bytes(&self.pk, <char *>b"parent_id", 9)
-                    if ret != 0: return ret
-                    ret = self._pack(o.parent_id)
-                    if ret != 0: return ret
-
-                    ret = pack_bytes(&self.pk, <char *>b"span_id", 7)
-                    if ret != 0: return ret
-                    ret = self._pack(o.span_id)
-                    if ret != 0: return ret
-
-                    ret = pack_bytes(&self.pk, <char *>b"service", 7)
-                    if ret != 0: return ret
-                    ret = self._pack(o.service)
-                    if ret != 0: return ret
-
-                    ret = pack_bytes(&self.pk, <char *>b"resource", 8)
-                    if ret != 0: return ret
-                    ret = self._pack(o.resource)
-                    if ret != 0: return ret
-
-                    ret = pack_bytes(&self.pk, <char *>b"name", 4)
-                    if ret != 0: return ret
-                    ret = self._pack(o.name)
-                    if ret != 0: return ret
-
-                    ret = pack_bytes(&self.pk, <char *>b"error", 5)
-                    if ret != 0: return ret
-                    ret = self._pack(1 if o.error else 0)
-                    if ret != 0: return ret
-
-                    ret = pack_bytes(&self.pk, <char *>b"start", 5)
-                    if ret != 0: return ret
-                    ret = self._pack(o.start_ns)
-                    if ret != 0: return ret
-
-                    ret = pack_bytes(&self.pk, <char *>b"duration", 8)
-                    if ret != 0: return ret
-                    ret = self._pack(o.duration_ns)
-                    if ret != 0: return ret
-
-                    if has_span_type:
-                        ret = pack_bytes(&self.pk, <char *>b"type", 4)
-                        if ret != 0: return ret
-                        ret = self._pack(o.span_type)
-                        if ret != 0: return ret
-
-                    if has_meta:
-                        ret = pack_bytes(&self.pk, <char *>b"meta", 4)
-                        if ret != 0: return ret
-                        ret = self._pack(o.meta)
-                        if ret != 0: return ret
-
-                    if has_metrics:
-                        ret = pack_bytes(&self.pk, <char *>b"metrics", 7)
-                        if ret != 0: return ret
-                        ret = self._pack(o.metrics)
-                        if ret != 0: return ret
-            else:
-                PyErr_Format(TypeError, b"can not serialize '%.200s' object", Py_TYPE(o).tp_name)
+        if metrics is None:
+            ret = msgpack_pack_nil(&self.pk)
```

Comment:
Same here as above, this shouldn't ever be `None` based on the checks we have in `_pack_span`